ui/Model.elm

```diff
@@ -126,42 +154,113 @@ getReleaseList =
         )
         releases
 
 
+processFilterVersions : FilterSet -> List Release -> List Release
+processFilterVersions filterSet releases =
+    List.filter
+        (\{ details } ->
+            List.any
+                (\f -> String.startsWith f details.filename)
+                (Dict.filter (\k v -> v) filterSet |> Dict.keys)
+        )
+        releases
+
+
+filterReleases : Model -> List Release
+filterReleases { filters, releases } =
+    releases
+        |> processFilter filters.channels (.details >> .channel)
+        |> processFilter filters.langs (.details >> .lang)
+        |> processFilter filters.targets (.details >> .target)
+        |> processFilterVersions filters.versions
+
+
 update : Msg -> Model -> ( Model, Cmd Msg )
-update message model =
+update message ({ filters } as model) =
     case message of
         ReleasesFetched result ->
             case result of
                 Ok releases ->
                     { model
                         | releases = releases
-                        , filters = { channels = extractChannels releases }
+                        , filters =
+                            { channels = extractFilterSet (.details >> .channel) releases
+                            , langs = extractFilterSet (.details >> .lang) releases
+                            , targets = extractFilterSet (.details >> .target) releases
+                            , versions = extractVersions releases
+                            }
                     }
                         ! []
 
                 Err err ->
                     Debug.crash "Unhandled Kinto error"
 
         ToggleChannelFilter channel active ->
-            { model
-                | filters =
-                    { channels =
-                        Dict.update channel (\_ -> Just active) model.filters.channels
+            let
+                channels =
+                    Dict.update channel (\_ -> Just active) model.filters.channels
+
+                newFilters =
+                    { filters | channels = channels }
+            in
+                { model | filters = newFilters } ! []
+
+        ToggleLangFilter lang active ->
+            let
+                newFilters =
+                    { filters
+                        | langs = Dict.update lang (\_ -> Just active) filters.langs
+                    }
+            in
+                { model | filters = newFilters } ! []
+
+        ToggleTargetFilter target active ->
+            let
+                newFilters =
+                    { filters
+                        | targets = Dict.update target (\_ -> Just active) filters.targets
+                    }
+            in
+                { model | filters = newFilters } ! []
+
+        ToggleVersionFilter version active ->
+            let
+                newFilters =
+                    { filters
+                        | versions = Dict.update version (\_ -> Just active) filters.versions
                     }
```

Comment:
These are all quite repetitive, why not extract them using a higher-order function?